File: ccwap/frontend/src/api/cost.ts
import { useQuery } from '@tanstack/react-query'
import { apiFetch, buildQuery } from './client'
import { costKeys } from './keys'
import type { DateRange } from '@/hooks/useDateRange'

export interface CostSummary {
  total_cost: number
  avg_daily_cost: number
  cost_today: number
  cost_this_week: number
  cost_this_month: number
  projected_monthly: number
}

export interface CostByTokenType {
  input_cost: number
  output_cost: number
  cache_read_cost: number
  cache_write_cost: number
  total_cost: number
}

export interface CostByModel {
  model: string
  cost: number
  turns: number
  input_tokens: number
  output_tokens: number
  percentage: number
}

export interface CostTrendPoint {
  date: string
  cost: number
  cumulative_cost: number
}

export interface CostByProject {
  project_path: string
  project_display: string
  cost: number
  percentage: number
}

export interface CacheSavings {
  total_cache_read_tokens: number
  total_input_tokens: number
  cache_hit_rate: number
  estimated_savings: number
  cost_without_cache: number
  actual_cost: number
}

export interface SpendForecast {
  daily_avg: number
  projected_7d: number
  projected_14d: number
  projected_30d: number
  trend_direction: string
  confidence: number
}

export interface CostAnalysisData {
  summary: CostSummary
  by_token_type: CostByTokenType
  by_model: CostByModel[]
  trend: CostTrendPoint[]
  by_project: CostByProject[]
  cache_savings: CacheSavings
  forecast: SpendForecast
}

export function useCostAnalysis(dateRange: DateRange) {
  return useQuery({
    queryKey: costKeys.data(dateRange),
    queryFn: () => apiFetch<CostAnalysisData>(
      `/cost${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

// --- Phase 4 additions ---

export interface CostAnomaly {
  date: string
  cost: number
  is_anomaly: boolean
  threshold: number
}

export interface CumulativeCost {
  date: string
  daily_cost: number
  cumulative: number
}

export interface CacheSimulation {
  actual_cost: number
  actual_cache_rate: number
  simulated_cost: number
  simulated_cache_rate: number
  savings: number
}

export function useCostAnomalies(dateRange: DateRange) {
  return useQuery({
    queryKey: costKeys.anomalies(dateRange),
    queryFn: () => apiFetch<CostAnomaly[]>(
      `/cost/anomalies${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

export function useCumulativeCost(dateRange: DateRange) {
  return useQuery({
    queryKey: costKeys.cumulative(dateRange),
    queryFn: () => apiFetch<CumulativeCost[]>(
      `/cost/cumulative${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

export function useCacheSimulation(targetHitRate: number, dateRange: DateRange) {
  return useQuery({
    queryKey: costKeys.cacheSimulation(targetHitRate, dateRange),
    queryFn: () => apiFetch<CacheSimulation>(
      `/cost/cache-simulation${buildQuery({ target_hit_rate: targetHitRate, from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

File: ccwap/server/models/cost.py
"""Pydantic models for cost analysis API."""

from typing import List, Optional, Dict
from pydantic import BaseModel


class CostSummary(BaseModel):
    """Cost summary cards."""
    total_cost: float = 0.0
    avg_daily_cost: float = 0.0
    cost_today: float = 0.0
    cost_this_week: float = 0.0
    cost_this_month: float = 0.0
    projected_monthly: float = 0.0


class CostByTokenType(BaseModel):
    """Cost breakdown by token type."""
    input_cost: float = 0.0
    output_cost: float = 0.0
    cache_read_cost: float = 0.0
    cache_write_cost: float = 0.0
    total_cost: float = 0.0


class CostByModel(BaseModel):
    """Cost for a specific model."""
    model: str
    cost: float = 0.0
    turns: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    percentage: float = 0.0


class CostTrendPoint(BaseModel):
    """A data point in the cost trend."""
    date: str
    cost: float = 0.0
    cumulative_cost: float = 0.0


class CostByProject(BaseModel):
    """Cost summary per project."""
    project_path: str
    project_display: str
    cost: float = 0.0
    percentage: float = 0.0


class CacheSavings(BaseModel):
    """Cache savings analysis."""
    total_cache_read_tokens: int = 0
    total_input_tokens: int = 0
    cache_hit_rate: float = 0.0
    estimated_savings: float = 0.0
    cost_without_cache: float = 0.0
    actual_cost: float = 0.0


class SpendForecast(BaseModel):
    """Spend forecast data."""
    daily_avg: float = 0.0
    projected_7d: float = 0.0
    projected_14d: float = 0.0
    projected_30d: float = 0.0
    trend_direction: str = "stable"
    confidence: float = 0.0


class CostAnalysisResponse(BaseModel):
    """Complete cost analysis data."""
    summary: CostSummary
    by_token_type: CostByTokenType
    by_model: List[CostByModel]
    trend: List[CostTrendPoint]
    by_project: List[CostByProject]
    cache_savings: CacheSavings
    forecast: SpendForecast
